File: ChessApp/src/CSharp8/ChessApp_Internals/ConsoleUI_Internals/GameCmdHandler.cs
using System;
using static MoreLinq.Extensions.ForEachExtension;
using ChessApp.ChessApp_Internals;
using Infrastructure;
using ChessApp.ConsoleUI_Internals.Commands;

namespace ChessApp.ConsoleUI_Internals
{
    public class GameCmdHandler
    {
        private GameFlow GameFlow { get; }
        private IStorage Storage { get; }
        private IConsoleIO Console { get; }

        public GameCmdHandler(GameFlow gameFlow, IStorage storage, IConsoleIO consoleIO)
        {
            GameFlow = gameFlow;
            Storage = storage;
            Console = consoleIO;
        }

        internal void Execute(GameCommand gc)
        {
            if (gc is NewGame ng)
            {
                GameFlow.NewGame(ng.UIPlayerColor);
            }
            else if (gc is LoadGame lc)
            {
                var (h, c) = Storage.Load(lc.Name);
                GameFlow.StartFrom(h, c);
            }
            else if (gc is ListGames)
            {
                Console.WriteLine("Saved games:");
                Storage.GetNames().ForEach(n => Console.WriteLine(n));
            }
            else if (gc is SaveGame sc)
            {
                var history = GameFlow.History;
                if (history == null)
                    Console.WriteLine("Nothing to save yet");
                else
                    Storage.Save(sc.Name, history, GameFlow.PlayerAColor!.Value);
            }
            else if (gc is DeleteGame dc)
            {
                Storage.Delete(dc.Name);
            }
            else
            {
                throw new ArgumentOutOfRangeException(nameof(gc));
            }
        }
    }
}

File: ChessApp/src/RustMacros/src/chess_app/console_ui/game_cmd_handler.rs
pub mod interface {
    use std::{fmt::Display, error::Error};
    use crate::chess_app::console_ui::data::command::Command;

    #[derive(Clone, Debug)]
    pub struct CmdError(pub String);

    impl Error for CmdError {
    }

    impl Display for CmdError {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            write!(f, "{}", self.0)
        }
    } 

    hlcd::define! {
        interface GameCmdHandler {
            fn execute(&self, cmd: Command) -> Result<(), CmdError>;
        }
    }
}

pub mod component {
    use crate::{hlcd_infra::console_io_interface::*, chess_app::{storage_interface::{*, self}, game_flow::interface::*, console_ui::data::command::Command}};
    use super::interface::*;

    hlcd::define! {
        component GameCmdHandler {

            requires {
                console_io: ConsoleIO,
                game_flow: GameFlowControl,
                storage: Storage
            }

            provides { GameCmdHandler }

            impl GameCmdHandler {
                fn execute(&self, cmd: Command) -> Result<(), CmdError> {
                    match cmd {
                        Command::MakeTurn(_) => unreachable!(),
                        Command::Exit => unreachable!(),
                        Command::Help => unreachable!(),
                        Command::NewGame(c) => { 
                            self.game_flow_mut().new_game(c); 
                        },
                        Command::ListGames => {
                            let con = self.console_io();
                            for name in self.storage().list_saved_games().unwrap() {
                                con.write(format!("{name}\n").as_str());
                            }
                        },
                        Command::LoadGame(name) => {
                            let (gh, c) = self.storage().load_game(name.as_str())?;
                            self.game_flow_mut().start_from(gh, c); 
                        },
                        Command::DeleteGame(name) => {
                            self.storage().delete_game(&name)?
                        },
                        Command::SaveGame(name) => {
                            let flow = self.game_flow();
                            if let Some(gh) = flow.game_history() {
                                self.storage().save_game(gh, flow.player_a_color().unwrap(), name.as_str())?;
                                let con = self.console_io();
                                con.write(format!("Game saved to '{name}'...\n").as_str());
                            } else {
                                return Err(CmdError("No active game to save".to_string()))
                            }
                        },
                    }
                    Ok(())
                }
            }
        }
    }

    impl From<storage_interface::Error> for CmdError {
        fn from(e: storage_interface::Error) -> Self {
            CmdError(format!("I/O error: {}", e.0))
        }
    }
}